core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java

```diff
@@ -379,12 +395,32 @@ public void start()
                     return;
                 }
 
+                planningThread = currentThread();
+                stateMachine.getStateChange(PLANNING).addListener(() -> {
+                    if (stateMachine.getQueryState() == FAILED) {
+                        synchronized (this) {
+                            if (planningThread != null) {
+                                planningThread.interrupt();
+                            }
+                        }
+                    }
+                }, directExecutor());
+
                 PlanRoot plan = planQuery();
                 // DynamicFilterService needs plan for query to be registered.
                 // Query should be registered before dynamic filter suppliers are requested in distribution planning.
                 registerDynamicFilteringQuery(plan);
                 planDistribution(plan);
 
+                synchronized (this) {
+                    if (Thread.interrupted()) {
```

Comment:
This shouldn't be necessary. It should be taken care of by the thread pool.

Reply:
This thread does not go into pool immediately. This might not be a problem, but it seems cleaner to clear interrupted flag since we interrupt planning specifically. wdyt?

Reply:
If planning is interrupted before this line, the above calls (planQuery, registerDynamicFilteringQuery, planDistribution) will throw and this will not be reached. 

The only case where it can happen is if there is a race condition where the planning completes just about the time the timeout kicks in and interrupts the thread right after `planDistribution` returns or while it's in some part that doesn't get affected by the interrupt so it doesn't throw. It definitely deserves a comment, and the debug statement is not very useful. Also see my comment above on how I'd restructure this section.

Reply:
> If planning is interrupted before this line, the above calls (planQuery, registerDynamicFilteringQuery, planDistribution) will throw and this will not be reached.

That will only happen if these methods will throw interrupted exception on IO which might not be the case.

> Also see my comment above on how I'd restructure this section.

+1

Reply:
> That will only happen if these methods will throw interrupted exception on IO which might not be the case.

That should not happen unless the code can do something reasonable with a lack of result from that IO. If we're silently ignoring those interruptions and making results up, that's a bigger problem.